Add tests for `xportr_length()`

averissimo opened the pull request:
Related issues:
- #82

🚨 both this and #87 add `metacore` :package: to `Suggests` section of the `DESCRIPTION` file. This needs to be changed in one of the PRs, or I can just change the base branch to #87 instead of `devel` _(after rebasing)_

## Changes description

* Add test cases to cover the `xportr_length()`
* Add `metacore` dependency to Suggests _(needs to address a condition where a )_
* Add `test/testthat/helper-length.R` [helper files](https://testthat.r-lib.org/reference/test_dir.html#special-files) to contain data and functions that are reused in tests

![image](https://user-images.githubusercontent.com/211358/234417722-35f378ca-1f10-463c-871a-25c72fe362c1.png)

averissimo commented on tests/testthat/test-length.R:
```diff
@@ -1,15 +1,209 @@
+#' * Result of call will create `SASlength` attribute (`width` for each
+#' variable)
+#'
+#' Helper functions / data (from ´test/testthat/helper-length.R´):
+#'  * \code{minimal_adsl} : minimal data frame with ADSL sample
+#'  * \code{minimal_length_metadata} : minimal metadata used in
+#'    `xport_length()`. It contains dataset (domain), variable names and length
+#'  * \code{expect_attr_width(result, metadata_length)} : support function to
+#'    test the data.frame modifications being done in `xportr_length()`
```
Sounds good to me! 

@bms63 Are you envisioning a "minimal metadata" and "minimal dataset" and including it in the scope of this PR?

Or a function that takes a bunch of logical parameters and builds that minimal example (`length`, `label`, `type`, ...)?

elimillera replied:
I know some other packages have some minimal data in their `data/` folder to use for testing and examples. I think that would be a great improvement to the testing and examples instead of using slightly different everything.

averissimo commented on tests/testthat/helper-data.R:
```diff
@@ -0,0 +1,55 @@
+# Re-usable data for sample ADSL table
```
I've changed the variables names to follow convention of other tests, so that it can be easier integrated in the future.

averissimo commented on tests/testthat/helper-mock_bindings.R:
```diff
@@ -0,0 +1,4 @@
+# function that can be used with mockery::stub to mock cli messages
```
The inclusion of `mockery` :package:  is purely cosmetical so that the output of running the tests doesn't have a bunch of empty lines due to the messages generated by `cli` _(margin on top for `cli_h2`)_.

I can remove it from both PRs if you think this is unnecessary

elimillera replied:
I think we would be good to actually change the tests/functions so they don't output lines instead of importing a new package. Too many imports raises issues of dependency risk.

averissimo replied:
done!

averissimo commented:
Updated the PR with comments:

* `minimal_{table,metadata}` are now functions that live in `R/` _(documented and tested)_
* Renamed tests to follow guidelines
* Removed all `helper-*.R` files in favor of using package functions
* Removed `mockery` from dependencies and forced the `cli` package to use theme within the tests
  * check code below on how to use it.

```
  # Remove empty lines in cli theme
  withr::local_options(list(cli.user_theme = cli_theme_tests))
  app <- cli::start_app(output = "message", .auto_close = FALSE)
  withr::defer(cli::stop_app(app))
```

This will prevent empty lines and allow us to catch all messages. Right now working for the tests in this PR, but can be used in other tests in the future.

bms63 commented on R/support-test.R:
```diff
@@ -0,0 +1,106 @@
+  )[cols_logical]
+
+  metadata <- tribble(
+    ~dataset, ~variable,        ~label, ~length,       ~type,       ~format, ~common,
```
Just wondering what is common variable?

bms63 replied:
Could we add an order column as well please.  Might one to add one or two more, but we can update later.

![image](https://user-images.githubusercontent.com/10111024/234374763-07c03c3a-40b8-4384-a944-8f1f32301ff9.png)

averissimo replied:
Oopsie. I must have meant another column that was used elsewhere and by mistake wrote `common`.

Thanks for catching this! I've replaced that with `order`.

The columns/data can always be revisited in the future!

bms63 commented on tests/testthat/test-length.R:
```diff
@@ -1,15 +1,214 @@
+
+
+test_that("xportr_length: Accepts valid domain names in metadata/metacore", {
+  adsl <- minimal_table(30)
+  metadata <- minimal_metadata(dataset = TRUE, length = TRUE, var_names = colnames(adsl))
```
This is amazing

bms63 commented on tests/testthat/test-length.R:
```diff
@@ -1,15 +1,214 @@
+#' variable)
+
+
+test_that("xportr_length: Accepts valid domain names in metadata/metacore", {
```
```suggestion
test_that("xportr_length: Accepts valid domain names in metadata object", {
```
Trying to keep generic.  @elimillera what d you think

bms63 replied:
Maybe just `metadata or metacore object`

averissimo replied:
That sounds good to me. There are no `metacore` references anywhere else in the documentation

bms63 commented on tests/testthat/test-support-for-tests.R:
```diff
@@ -0,0 +1,32 @@
+test_that("minimal_table: builds minimal data frame with data", {
```
Love the tests for the tester helpers!!

bms63 reviewed (changes requested):
I think an order column is need for the metadata.  After that this all looks good to me!

averissimo commented:
@bms63 I think I addressed all your comments 🕵🏽‍♀️ Let me know if you have any additional concerns

bms63 reviewed (approved):
Looks great!! Many thanks
